File: Solution/CSharp/101.SymmetricTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if (root == null) return true;
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while (queue.Count > 0) {
            int count = queue.Count;
            List<int?> nextLevel = new List<int?>();
            for (int i = 0; i < count; i++) {
                TreeNode curNode = queue.Dequeue();
                if (curNode.left != null) {
                    queue.Enqueue(curNode.left);
                    nextLevel.Add(curNode.left.val);
                } else {
                    nextLevel.Add(null);
                }
                if (curNode.right != null) {
                    queue.Enqueue(curNode.right);
                    nextLevel.Add(curNode.right.val);
                } else {
                    nextLevel.Add(null);
                }
            }
            if (!LevelIsSymmetric(nextLevel)) {
                return false;
            }
        }
        return true;

        public bool LevelIsSymmetric(List<int?> list) {
            int left = 0;
            int right = list.Count - 1;
            while (right > left) {
                if (list[left++] != list[right--]) {
                    return false;
                }
            }
            return true;
        }
    }

File: Solution/Python/101. Symmetric Tree.py
def getHelperPath():
    path = __file__.split("\\")[0:-1]
    path[-2] = "Helper"
    return "\\".join(path)


import sys
sys.path.append(getHelperPath())
from Tree import *

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        queue = [root]
        while queue and queue[0]:
            level = []
            count = len(queue)
            for i in range(count):
                node = queue.pop(0)

                if node.left:
                    queue.append(node.left)
                    level.append(node.left.val)
                else:
                    level.append(None)

                if node.right:
                    queue.append(node.right)
                    level.append(node.right.val)
                else:
                    level.append(None)

            if level != level[::-1]:
                return False

        return True


root = Tree.CreateTree([1, 2, 2, 3, 4, 4, 3])
s = Solution()
res = s.isSymmetric(root)
print(res)
